master/internal/api_trials.go

```diff
@@ -419,3 +409,141 @@ func (a *apiServer) GetTrial(_ context.Context, req *apiv1.GetTrialRequest) (
+	resp apiv1.Determined_GetTrialProfilerMetricsServer,
+) error {
+	switch exists, err := a.m.db.CheckTrialExists(int(req.Labels.TrialId)); {
+	case err != nil:
+		return err
+	case !exists:
+		return status.Error(codes.NotFound, "trial not found")
+	}
+
+	labelsParam, err := protojson.Marshal(req.Labels)
+	if err != nil {
+		return fmt.Errorf("failed to marshal labels: %w", err)
+	}
+
+	fetch := func(lr api.LogsRequest) (api.LogBatch, error) {
+		switch {
+		case lr.Follow, lr.Limit > trialProfilerMetricsBatchSize:
+			lr.Limit = trialProfilerMetricsBatchSize
+		case lr.Limit <= 0:
+			return nil, nil
+		}
+
+		var batchOfBatches []*trialv1.TrialProfilerMetricsBatch
+		return model.TrialProfilerMetricsBatchBatch(batchOfBatches), a.m.db.QueryProto(
+			"get_trial_profiler_metrics",
+			&batchOfBatches, labelsParam, lr.Offset, lr.Limit,
+		)
+	}
+
+	onBatch := func(b api.LogBatch) error {
+		return b.ForEach(func(r interface{}) error {
+			return resp.Send(&apiv1.GetTrialProfilerMetricsResponse{
+				Batch: r.(*trialv1.TrialProfilerMetricsBatch),
+			})
+		})
+	}
+
+	return a.m.system.MustActorOf(
+		actor.Addr("trialMetricsStream-"+uuid.New().String()),
+		api.NewLogStoreProcessor(
+			resp.Context(),
+			api.LogsRequest{Follow: true},
+			fetch,
+			onBatch,
+			a.isTrialTerminalFunc(int(req.Labels.TrialId), -1),
+			&defaultTrialProfilerMetricsBatchWaitTime,
+		),
+	).AwaitTermination()
+}
+
+func (a *apiServer) GetTrialProfilerAvailableSeries(
+	req *apiv1.GetTrialProfilerAvailableSeriesRequest,
+	resp apiv1.Determined_GetTrialProfilerAvailableSeriesServer,
+) error {
+	switch exists, err := a.m.db.CheckTrialExists(int(req.TrialId)); {
+	case err != nil:
+		return err
+	case !exists:
+		return status.Error(codes.NotFound, "trial not found")
+	}
+
+	fetch := func(_ api.LogsRequest) (api.LogBatch, error) {
+		var labels apiv1.GetTrialProfilerAvailableSeriesResponse
+		return api.ToLogBatchOfOne(&labels), a.m.db.QueryProto(
+			"get_trial_available_series",
+			&labels, fmt.Sprintf(`{"trialId": %d}`, req.TrialId),
+		)
+	}
+
+	onBatch := func(b api.LogBatch) error {
+		return b.ForEach(func(r interface{}) error {
+			return resp.Send(r.(*apiv1.GetTrialProfilerAvailableSeriesResponse))
+		})
+	}
+
+	return a.m.system.MustActorOf(
+		actor.Addr("trialSeriesStream-"+uuid.New().String()),
+		api.NewLogStoreProcessor(
+			resp.Context(),
+			api.LogsRequest{Follow: true},
+			fetch,
+			onBatch,
+			nil,
+			&DefaultTrialAvailableSeriesBatchWaitTime,
+		),
+	).AwaitTermination()
+}
+
+func (a *apiServer) PostTrialProfilerMetricsBatch(
+	_ context.Context,
+	req *apiv1.PostTrialProfilerMetricsBatchRequest,
+) (*apiv1.PostTrialProfilerMetricsBatchResponse, error) {
+	switch exists, err := a.m.db.CheckTrialExists(int(req.Batch.Labels.TrialId)); {
+	case err != nil:
+		return nil, err
+	case !exists:
+		return nil, status.Error(codes.NotFound, "trial not found")
+	}
+
+	if len(req.Batch.Values) != len(req.Batch.Batches) &&
```

Comment:
If we are trying to detect if the arrays differ in size at all, should this be an OR rather than an AND?